OrchardZSA backward compatability 0.8.0

dmidem opened the pull request:
This pull request aims to generalize the implementation of the Orchard protocol, providing  backward compatibility to support both the new (ZSA) and the original (non-ZSA - Vanilla) Orchard protocol variants. Key modifications and enhancements include:

### 1. Trait and Generic Structures for Note Encryption:
- **OrchardDomain Trait:** A new `OrchardDomain` trait in `note_encryption.rs` differentiates between the original Orchard (Vanilla) and Orchard ZSA protocols, simplifying the implementation of the `Domain` trait through abstraction.
- **OrchardDomainBase Generic Struct:** Introduced to contain data used for internal calculations in both Orchard variants.
- **TransmittedNoteCiphertext Modifications:** This struct is now generic, supporting various lengths of encrypted note ciphertext to accommodate both Orchard variants.

### 2. Trait and Generic Structures for Circuit Generalization:
- **OrchardCircuit Trait:** A new `OrchardCircuit` trait in `circuit.rs` provides an interface for different implementations of the PLONK circuit tailored to the specific requirements of the Orchard protocol's variants (Vanilla and ZSA).
- **OrchardCircuitBase Generic Struct:** Contains data for internal calculations across both Orchard variants.

### 3. Module Organization:
- Introduction of `note_encryption_vanilla.rs`, `note_encryption_zsa.rs` to support the different types of encrypted notes.
- Introduction of `circuit_vanilla.rs`, and `circuit_zsa.rs` to support various circuit configurations.

### 4. Test Suite Updates:
Updates to unit tests include separate versions for Vanilla and ZSA variants, ensuring thorough validation of the modifications.

### 5. Dependency Adaptation:
The adoption of a modified version of the Halo2 Rust crate facilitates support for both Orchard protocol variants, guaranteeing that all tests, including those for non-ZSA functionality, pass successfully.

PaulLaux commented on src/bundle/commitments.rs:
```diff
@@ -28,12 +70,21 @@ fn hasher(personal: &[u8; 16]) -> State {
 ///
-/// Then, hash these together along with (flags, value_balance_orchard, anchor_orchard),
-/// personalized with ZCASH_ORCHARD_ACTIONS_HASH_PERSONALIZATION
+/// These hashes are combined with additional data, which may include the ZSA burn field for specific
+/// Orchard variants (see ZIP-226), flags, value_balance_orchard, and anchor_orchard. The inclusion of
+/// the burn field is dependent on the specific implementation of the `OrchardHash` trait, which
+/// defines whether the burn field should be included based on the variant (ZSA variant includes it).
```
```rust
/// Then, hash these together along with (flags, value_balance_orchard, anchor_orchard),
/// and potentially the burn fields, if it is an OrchardZSA action.
```

dmidem replied:
Fixed.

PaulLaux commented on src/bundle/commitments.rs:
```diff
@@ -19,6 +27,40 @@ fn hasher(personal: &[u8; 16]) -> State {
     Params::new().hash_length(32).personal(personal).to_state()
 }
 
+// FIXME: Consider not using a separate OrchardHash trait and instead move update_hash_with_burn to
+// the OrchardDomain or OrchardFlavour trait.
```
yes, that is too much for one conditional update. let's move it to OrchardDomain or OrchardFlavour

dmidem replied:
Done: used OrchardFlavor instead of OrchardDomainCommon+OrchardCircuit+OrchardHash and renamed the corresponding generic arg from D to FL.

PaulLaux commented on src/note_encryption/action.rs:
```diff
@@ -0,0 +1,162 @@
+
+    use super::{CompactAction, OrchardDomain, OrchardDomainBase};
+
+    // FIXME: pin fake_compact_action function to OrchardZSA or keep it generic?
```
let's keep it generic

PaulLaux commented on src/note_encryption/action.rs:
```diff
@@ -0,0 +1,162 @@
+                }
+            }
+        };
+        // FIXME: consider using another AssetBase instead of native for ZSA.
```
no need

dmidem replied:
Got it, removed the comment.

PaulLaux commented on src/note_encryption/domain.rs:
```diff
@@ -0,0 +1,173 @@
+//! Domain-specific requirements for Orchard note encryption and decryption.
+//!
+//! This module implements Domain and BatchDomain traits from zcash_note_encryption
+//! crate for OrchardDomain types, i.e. contains a common logic of both Vanilla and
+//! ZSA variations of the Orchard protocol.
```
```rust
//! This module implements Domain and BatchDomain traits from the zcash_note_encryption
//! crate and contains the common logic for `OrchardVanilla` and `OrchardZSA` flavors.
```

dmidem replied:
Fixed.

PaulLaux commented on src/note_encryption/orchard_domain.rs:
```diff
@@ -0,0 +1,79 @@
+//! Represents the Orchard protocol domain specifics required for note encryption and decryption.
+//!
+//! Defines OrchardDomain triat with the constants and methods needed to handle the
+//! various sizes and types of note components. Actually this trait represents the difference
+//! between both Vanilla and ZSA variantions of the Orchard protocol.
```
//! The OrchardDomain trait represents the difference between the Orchard and the OrchardZSA 
//! encryption and decryption procedures.

dmidem replied:
Fixed.

PaulLaux commented on src/note_encryption/orchard_domain.rs:
```diff
@@ -0,0 +1,79 @@
+/// Orchard-specific note encryption logic.
+#[derive(Debug, Clone)]
+pub struct OrchardDomainBase<D: OrchardDomain> {
+    /// Represents a nullifier which is used to prevent double spending within the Orchard protocol.
```
remove comment, no need for global statements. 

Aside, Rho is not a nullifier but one of the parameters needed to generate the nullifier.

dmidem replied:
As we discussed, we need to keep it as it's a doc comment required by Rust for the pub structs. I have updated the comment to reflect the proper meaning.

PaulLaux reviewed (commented):
Very good overall, added comments, mostly about the note_encryption structure.
The structure need to reflect the generalization
```
Domain -> OrchardDomain 
                       |-> OrchardDomainVanilla
                       |-> OrchardDomainZSA
```
Which is currently not the case.

In addition,
 - src/circuit/circuit_description_vanilla should be represented as a moved file in the diff, not a new file

dmidem commented on src/note_encryption/note_bytes.rs:
```diff
@@ -0,0 +1,48 @@
+}
+
+/// Defines the ability to concatenate two byte slices.
+pub trait NoteByteConcat: for<'a> From<(&'a [u8], &'a [u8])> {}
+
+impl<const N: usize> NoteByteConcat for NoteBytes<N> {}
+
+/// Defines the behavior for types that can provide read-only access to their internal byte array.
+pub trait NoteByteReader: AsRef<[u8]> + for<'a> From<&'a [u8]> + Clone + Copy {}
+
+impl<const N: usize> NoteByteReader for NoteBytes<N> {}
+
+/// Defines the behavior for types that support both read and write access to their internal byte array.
+pub trait NoteByteWriter: NoteByteReader + AsMut<[u8]> {}
+
+impl<const N: usize> NoteByteWriter for NoteBytes<N> {}
```
Done.

dmidem commented on src/note_encryption/orchard_domain.rs:
```diff
@@ -0,0 +1,79 @@
+/// Represents the Orchard protocol domain specifics required for note encryption and decryption.
+pub trait OrchardDomain: fmt::Debug + Clone {
+    /// The size of a compact note, specific to the Orchard protocol.
+    // 52 for Vanuilla, 84 for ZSA
```
Done.

dmidem commented on src/note_encryption/orchard_domain.rs:
```diff
@@ -0,0 +1,79 @@
+    const COMPACT_NOTE_SIZE: usize;
+
+    /// The size of a note plaintext, including memo and other metadata.
+    // + 512 (564 for Vanilla, 596 for ZSA)
```
Done.

dmidem commented on src/note_encryption/orchard_domain.rs:
```diff
@@ -0,0 +1,79 @@
+    const NOTE_PLAINTEXT_SIZE: usize = Self::COMPACT_NOTE_SIZE + MEMO_SIZE;
+
+    /// The size of an encrypted note ciphertext, accounting for additional AEAD tag space.
+    // + 16 (580 for Vanilla, 612 for ZSA)
```
Done.

dmidem commented on src/note_encryption/orchard_domain.rs:
```diff
@@ -0,0 +1,79 @@
+    // + 16 (580 for Vanilla, 612 for ZSA)
+    const ENC_CIPHERTEXT_SIZE: usize = Self::NOTE_PLAINTEXT_SIZE + AEAD_TAG_SIZE;
+
+    /// A type to represent the raw bytes of a note plaintext.
```
Done.

dmidem commented on src/note_encryption/action.rs:
```diff
@@ -0,0 +1,162 @@
+        Some(self.encrypted_note().enc_ciphertext)
+    }
+
+    // FIXME: split at COMPACT_NOTE_SIZE - is this correct?
```
Got it. Removed the comment.

dmidem commented on src/note_encryption/note_encryption_zsa.rs:
```diff
@@ -0,0 +1,212 @@
+//! In-band secret distribution for Orchard bundles (ZSA version).
+//!
+//! This module implements the note encryption logic specific to the Zcash shielded assets (ZSA)
```
Fixed.

dmidem commented on src/note_encryption/note_encryption_vanilla.rs:
```diff
@@ -0,0 +1,211 @@
+//! In-band secret distribution for Orchard bundles (Vanilla variation).
+//!
+//! This module implements the note encryption logic specific to the standard ("Vanilla")
```
Fixed.

dmidem commented on src/builder.rs:
```diff
@@ -700,14 +715,16 @@ fn pad_spend(spend: Option<&SpendInfo>, asset: AssetBase, mut rng: impl RngCore)
 /// The returned bundle will have no proof or signatures; these can be applied with
 /// [`Bundle::create_proof`] and [`Bundle::apply_signatures`] respectively.
-pub fn bundle<V: TryFrom<i64>>(
+// FIXME: Consider factoring parts of the return type into `type` definitions
```
Got it, removed the comment.

dmidem commented on src/builder.rs:
```diff
@@ -620,10 +633,12 @@ impl Builder {
     /// The returned bundle will have no proof or signatures; these can be applied with
     /// [`Bundle::create_proof`] and [`Bundle::apply_signatures`] respectively.
-    pub fn build<V: TryFrom<i64>>(
+    // FIXME: Consider factoring parts of the return type into `type` definitions
```
Neve mind, I removed the comment (see the following issue).

dmidem commented on src/note_encryption/note_encryption_zsa.rs:
```diff
@@ -0,0 +1,212 @@
+
+    proptest! {
+        #[test]
+        fn test_encoding_roundtrip(
```
Done (checked with `cargo test test_encoding_roundtrip` - it really runs test_encoding_roundtrip test for vanilla and zsa).

dmidem commented on src/note_encryption/note_encryption_vanilla.rs:
```diff
@@ -0,0 +1,211 @@
+//! In-band secret distribution for Orchard bundles (Vanilla variation).
```
Renamed.

dmidem commented on src/note_encryption/note_encryption_zsa.rs:
```diff
@@ -0,0 +1,212 @@
+//! In-band secret distribution for Orchard bundles (ZSA version).
```
Renamed.

dmidem commented on src/builder.rs:
```diff
@@ -425,15 +438,15 @@ impl ActionInfo {
                 },
             ),
-            Circuit::from_action_context_unchecked(self.spend, note, alpha, self.rcv),
+            CircuitBase::<D>::from_action_context_unchecked(self.spend, note, alpha, self.rcv),
```
Renamed CircuitBase back to Circuit.

dmidem commented:
> Very good overall, added comments, mostly about the note_encryption structure. The structure need to reflect the generalization
> 
> ```
> Domain -> OrchardDomain 
>                        |-> OrchardDomainVanilla
>                        |-> OrchardDomainZSA
> ```
> 
> Which is currently not the case.
> 
> In addition,
> 
> * src/circuit/circuit_description_vanilla should be represented as a moved file in the diff, not a new file

As I've created the `orchardzsa-backward-compatability-0.8.0` branch from the `zsa1` branch, `src/circuit/circuit_description_vanilla` and `src/circuit/circuit_proof_test_case_vanilla.bin` are indeed new files (copied from `zcash/orchard` upstream branch). However, `src/circuit/circuit_description_zsa` and `src/circuit/circuit_proof_test_case_zsa.bin` are renamed files from `zsa1`.

Originally, `src/circuit/circuit_description_zsa` was correctly marked as renamed, but `src/circuit/circuit_proof_test_case_zsa.bin` was mistakenly shown as new because I removed `src/circuit_proof_test_case.bin` in one commit and added `src/circuit/circuit_proof_test_case_zsa.bin` in a later commit. I have fixed this by restoring the original `src/circuit_proof_test_case.bin`, deleting `src/circuit/circuit_proof_test_case_zsa.bin`, and then performing a proper `git mv src/circuit_proof_test_case.bin src/circuit/circuit_proof_test_case_zsa.bin`. It now shows correctly as a renamed file in the PR diff.

dmidem commented on src/note_encryption.rs:
```diff
@@ -1,108 +1,89 @@
-    }
-}
+/// The size of the encoding of a ZSA asset id.
+const ZSA_ASSET_SIZE: usize = 32;
```
Done: moved the content to `note_encryption/domain.rs` and `note_encryption/orchard_domain.rs`

dmidem commented on src/note_encryption/note_bytes.rs:
```diff
@@ -0,0 +1,48 @@
+//! Utilities for handling note bytes in the Orchard protocol.
```
Moved the content to `note_encryption/orchard_domain.rs`.

PaulLaux commented on src/bundle.rs:
```diff
@@ -272,25 +281,65 @@ impl<T: Authorization, V> Bundle<T, V> {
+    // converting types. However, the types for burn values are now always NoteValue.
+    //
+    // WARNING: Contains unwrap() calls! DO NOT remove #[cfg(test)] or handle unwrap() carefully!
+    #[cfg(test)]
```
Maybe better to move to test section?

dmidem replied:
Removed `#[cfg(test)]` since `try_map_value_balance` is now simplified. As discussed, we've decided to keep `try_map_value_balance` in its original place.

PaulLaux commented on src/bundle/commitments.rs:
```diff
@@ -44,19 +81,28 @@ pub(crate) fn hash_bundle_txid_data<A: Authorization, V: Copy + Into<i64>>(
     h.update(mh.finalize().as_bytes());
     h.update(nh.finalize().as_bytes());
+
+    // Delegate complete handling of the burn data to the OrchardHash implementation
```
remove comment

dmidem replied:
Removed.

PaulLaux commented on src/note_encryption/domain.rs:
```diff
@@ -0,0 +1,298 @@
+pub(super) const COMPACT_NOTE_SIZE_ZSA: usize = COMPACT_NOTE_SIZE_VANILLA + ZSA_ASSET_SIZE;
+
+/// The version byte for Vanilla.
+pub(super) const NOTE_VERSION_BYTE_VANILLA: u8 = 0x02;
```
`NOTE_VERSION_BYTE_V2`

dmidem replied:
Renamed.

PaulLaux commented on src/note_encryption/domain.rs:
```diff
@@ -0,0 +1,298 @@
+pub(super) const NOTE_VERSION_BYTE_VANILLA: u8 = 0x02;
+
+/// The version byte for ZSA.
+pub(super) const NOTE_VERSION_BYTE_ZSA: u8 = 0x03;
```
`NOTE_VERSION_BYTE_V3`

dmidem replied:
Renamed.

PaulLaux commented on src/note_encryption/orchard_domain.rs:
```diff
@@ -0,0 +1,113 @@
+    }
+}
+
+// FIXME: consider implementing and using TryFrom instead (see also review comments in
+// zcash_note_encryption PR)
```
remove comment. We have it under a different task

dmidem replied:
Removed.

PaulLaux commented on src/orchard_flavors.rs:
```diff
@@ -0,0 +1,18 @@
+//! Defines types and traits for the variations ("flavors") of the Orchard protocol (Vanilla and ZSA).
```
rename file to `orchard_flavor.rs`

dmidem replied:
Renamed.

PaulLaux commented on src/orchard_flavors.rs:
```diff
@@ -0,0 +1,18 @@
+
+use crate::{bundle::OrchardHash, circuit::OrchardCircuit, note_encryption::OrchardDomainCommon};
+
+/// Represents the standard ("Vanilla") variation ("flavor") of the Orchard protocol.
```
`Represents the "Vanilla" variation ("flavor") of the Orchard protocol.`

dmidem replied:
Fixed.

PaulLaux commented on src/orchard_flavors.rs:
```diff
@@ -0,0 +1,18 @@
+#[derive(Debug, Clone, Default)]
+pub struct OrchardZSA;
+
+/// A trait binding the common functionality between different Orchard protocol variations
```
`A trait binding the common functionality between different Orchard protocol flavors.`

dmidem replied:
Fixed.

PaulLaux reviewed (commented):
Added some relatively minor comments. In addition:

- in `benches/..` please parametrize every **benchmark group** label with `Orchard` or `OrchardZSA` according to the relevant case. For example in `benched/note_encryption` we need:
```rust
let mut group = c.benchmark_group("note-decryption");
```
to be replaced with `Orchard note-decryption` and `OrchardZSA note-decryption` according to the performed case.
This is needed to properly display and separate the results.
see results using `cargo bench` and make sure all benchmark tests are executed.
same for `/circuit`. My advice is to avoid new generics and use functions.

dmidem commented on src/note_encryption/compact_action.rs:
```diff
@@ -1,12 +1,18 @@
 //! Defines actions for Orchard shielded outputs and compact action for light clients.
 
+// FIXME: move to upper (src) folder?
```
Got it. Removed the FIXME comment.

dmidem commented on src/note_encryption/orchard_domain_vanilla.rs:
```diff
@@ -76,7 +81,7 @@ mod tests {
             // Decode.
             let domain = OrchardDomainVanilla::for_rho(rho);
-            let parsed_version = note_version(plaintext.as_ref()).unwrap();
+            let parsed_version = parse_note_version(plaintext.as_ref()).unwrap();
```
Fixed.

dmidem commented on benches/note_decryption.rs:
```diff
@@ -157,11 +158,25 @@ fn bench_note_decryption(c: &mut Criterion) {
+    targets = bench_note_decryption::<OrchardZSA>
+}
+
+criterion_main!(benches_vanilla, benches_zsa);
```
I've moved the OS-specific logic into the `create_config` function, so we no longer need the `[cfg(not(unix))]` directive. The function handles the configuration based on the OS automatically.

dmidem commented on src/action.rs:
```diff
@@ -142,70 +145,90 @@ pub(crate) mod testing {
-                encrypted_note,
-                cv_net,
-                authorization: ()
+    /// `ActionArb` adapts `arb_...` functions for both Vanilla and ZSA Orchard protocol variations
+    /// in property-based testing, addressing proptest crate limitations.
```
Removed.

dmidem commented on src/action.rs:
```diff
@@ -142,70 +145,90 @@ pub(crate) mod testing {
+                );
+
+                let mut rng = StdRng::from_seed(rng_seed);
+                let encryptor = NoteEncryption::<OrchardDomain<D>>::new(None, note, memo.try_into().unwrap());
```
Done.

dmidem commented on src/bundle/batch.rs:
```diff
@@ -35,9 +38,9 @@ impl BatchValidator {
 
     /// Adds the proof and RedPallas signatures from the given bundle to the validator.
-    pub fn add_bundle<V: Copy + Into<i64>>(
+    pub fn add_bundle<V: Copy + Into<i64>, D: OrchardDomainCommon + OrchardHash>(
```
Done.

dmidem commented on src/builder.rs:
```diff
@@ -620,10 +635,11 @@ impl Builder {
     /// The returned bundle will have no proof or signatures; these can be applied with
     /// [`Bundle::create_proof`] and [`Bundle::apply_signatures`] respectively.
-    pub fn build<V: TryFrom<i64>>(
+    #[allow(clippy::type_complexity)]
```
Removed.

dmidem commented on src/builder.rs:
```diff
@@ -109,7 +113,9 @@ impl BundleType {
             BundleType::Transactional { flags, .. } => *flags,
-            BundleType::Coinbase => Flags::SPENDS_DISABLED,
+            // FIXME: is it correct to use SPENDS_DISABLED_WITHOUT_ZSA (i.e does coinbase always have
+            // ZSA disabled)?
```
Got it. Removed the comment.

dmidem commented on src/builder.rs:
```diff
@@ -825,20 +845,28 @@ pub fn bundle<V: TryFrom<i64>>(
-        burn.iter()
-            .flat_map(|(asset, value)| -> Result<_, BuildError> { Ok((*asset, (*value).into()?)) }),
+        burn.iter().cloned(),
+        //burn.iter().flat_map(|(asset, value)| -> Result<_, BuildError> { Ok((*asset, (*value).into()?)) }),
```
Done.

dmidem commented on src/builder.rs:
```diff
@@ -825,20 +845,28 @@ pub fn bundle<V: TryFrom<i64>>(
+                asset,
+                NoteValue::from_raw(
+                    u64::try_from(i128::from(value))
+                        .map_err(|_| BuildError::ValueSum(value::OverflowError))?,
```
Done.

dmidem commented on src/builder.rs:
```diff
@@ -870,6 +898,19 @@ pub struct InProgress<P, S: InProgressSignatures> {
 }
 
+impl<P, S: InProgressSignatures> InProgress<P, S> {
+    /// Changes this authorization from one proof type to another.
```
This came from @alexeykoren's commit 5ea48b46. It seems to be needed for converting the proof type for use on the librustzcash side, but I can't find where it's specifically used.

dmidem commented on src/builder.rs:
```diff
@@ -894,13 +935,15 @@ impl<S: InProgressSignatures> InProgress<Unproven, S> {
 }
 
-impl<S: InProgressSignatures, V> Bundle<InProgress<Unproven, S>, V> {
+impl<S: InProgressSignatures, V, D: OrchardDomainCommon + OrchardCircuit>
```
Done.

dmidem commented on src/bundle.rs:
```diff
@@ -185,28 +193,29 @@ pub trait Authorization: fmt::Debug {
     value_balance: V,
     /// Assets intended for burning
-    burn: Vec<(AssetBase, V)>,
+    // FIXME: use BurnType like it's in Zebra? Put it as another param of Domain trait
```
Removed.

dmidem commented on tests/builder.rs:
```diff
@@ -47,24 +52,26 @@ pub fn build_merkle_path(note: &Note) -> (MerklePath, Anchor) {
-                flags: Flags::SPENDS_DISABLED,
+                // FIXME: Should we use SPENDS_DISABLED_WITH_ZSA for FL=OrchardZSA case?
+                // The original zsa1 code used the following:
+                // let mut builder = Builder::new(Flags::from_parts(false, true, false), anchor);
```
Got it - fixed.

dmidem commented on tests/builder.rs:
```diff
@@ -96,20 +103,44 @@ fn bundle_chain() {
+    let shielded_bundle: Bundle<_, i64, FL> = {
         let (merkle_path, anchor) = build_merkle_path(&note);
 
+        // FIXME: Should we use DEFAULT_ZSA for FL=OrchardZSA case?
```
Fixed.

dmidem commented on Cargo.toml:
```diff
@@ -26,11 +26,12 @@ rustdoc-args = ["--cfg", "docsrs", "--html-in-header", "katex-header.html"]
 group = { version = "0.13", features = ["wnaf-memuse"] }
-halo2_gadgets = { git = "https://github.com/QED-it/halo2", branch = "zsa1" }
-halo2_proofs = { git = "https://github.com/QED-it/halo2", branch = "zsa1", default-features = false, features = ["batch", "floor-planner-v1-legacy-pdqsort"] }
+halo2_gadgets = { git = "https://github.com/QED-it/halo2", branch = "improve-backward-compatability-with-zsa" }
+halo2_proofs = { git = "https://github.com/QED-it/halo2", branch = "improve-backward-compatability-with-zsa", default-features = false, features = ["batch", "floor-planner-v1-legacy-pdqsort"] }
```
Done.

dmidem commented on Cargo.toml:
```diff
@@ -56,14 +57,12 @@ plotters = { version = "0.3.0", optional = true }
-halo2_gadgets = { git = "https://github.com/QED-it/halo2", branch = "zsa1", features = ["test-dependencies"] }
-hex = "0.4"
+criterion = "0.4" # 0.5 depends on clap 4 which has MSRV 1.70
+halo2_gadgets = { git = "https://github.com/QED-it/halo2", branch = "improve-backward-compatability-with-zsa", features = ["test-dependencies"] }
```
Done.

dmidem commented on src/builder.rs:
```diff
@@ -700,14 +718,16 @@ fn pad_spend(spend: Option<&SpendInfo>, asset: AssetBase, mut rng: impl RngCore)
     bundle_type: BundleType,
     spends: Vec<SpendInfo>,
     outputs: Vec<OutputInfo>,
+    // FIXME: Should we use NoteValue instead of ValueSum here as well?
```
Done.

dmidem commented on src/bundle.rs:
```diff
@@ -272,25 +281,65 @@ impl<T: Authorization, V> Bundle<T, V> {
+    // the conversion. This function is restricted to test configuration due
+    // to the use of unwrap calls.
+    //
+    // An alternative approach could be to modify this function to convert only
+    // value_balance without converting burn values, as in practice, we are only
+    // converting types. However, the types for burn values are now always NoteValue.
```
Done (removed the FIXME comment completely as `try_map_value_balance` is now simplified).

dmidem commented on src/bundle.rs:
```diff
@@ -272,25 +281,65 @@ impl<T: Authorization, V> Bundle<T, V> {
+    // function to convert a ValueSum-based Bundle to an Amount-based Bundle.
+    //
+    // It converts balance and burn values from one type to another (from ValueSum to Amount).
+    // To achieve this, the function uses a complex series of `TryFrom` conversions
+    // and `unwrap` calls. These conversions and unwraps are necessary to convert
```
Done (removed the FIXME comment completely as `try_map_value_balance` is now simplified).

dmidem commented on src/bundle.rs:
```diff
@@ -272,25 +281,65 @@ impl<T: Authorization, V> Bundle<T, V> {
+
+    // FIXME:
+    //
+    // This function is currently used in the `testing` module of the
+    // `librustzcash/zcash_primitives` crate, specifically in the `arb_bundle`
+    // function to convert a ValueSum-based Bundle to an Amount-based Bundle.
```
Done - removed the FIXME comment completely as `try_map_value_balance` is now simplified as discussed: the function now converts only `value_balance` without converting `burn` values.

PaulLaux commented on benches/note_decryption.rs:
```diff
@@ -15,7 +15,11 @@ use zcash_note_encryption_zsa::{batch, try_compact_note_decryption, try_note_dec
 use pprof::criterion::{Output, PProfProfiler};
 
-fn bench_note_decryption<FL: OrchardFlavor>(c: &mut Criterion) {
+mod utils;
```
When running `cargo bench`, I see tests for proving and verifying for Orchard and OrchardZSA.
I don't see the decryption tests. Are you sure it is activated?

PaulLaux reviewed (commented):
Good, except benchmark for decryption does not run

PaulLaux commented on benches/note_decryption.rs:
```diff
@@ -28,6 +28,8 @@ fn bench_note_decryption<FL: OrchardFlavorBench>(c: &mut Criterion) {
     let valid_ivk = PreparedIncomingViewingKey::new(&valid_ivk);
 
+    // FIXME: should we add to the following comment that now for ZSA flavor the early
+    // rejection also happens when the asset is invalid?
```
no

dmidem replied:
Got it. Removed the comment.

PaulLaux commented on src/builder.rs:
```diff
@@ -584,7 +584,7 @@ impl Builder {
 
     /// Add an instruction to burn a given amount of a specific asset.
-    // FIXME: Should we add `add_burn` into `testing` (into `arb_bundle` function)
+    // FIXME: Should we add `add_burn` into the `into_bundle` function in this testing module
```
no, I opened a new task for later

dmidem replied:
Got it. Removed the comment.

PaulLaux commented on src/note_encryption/orchard_domain.rs:
```diff
@@ -62,6 +66,9 @@ pub trait OrchardDomainCommon: fmt::Debug + Clone {
     /// The size of an encrypted note ciphertext, accounting for additional AEAD tag space.
     const ENC_CIPHERTEXT_SIZE: usize = Self::NOTE_PLAINTEXT_SIZE + AEAD_TAG_SIZE;
 
+    /// The note version byte.
+    const NOTE_VERSION_BYTE: u8;
+
```
It is not working. we should not bake the version byte as part of the domain.

PaulLaux replied:
The version is a property of the note, not the encryption.

PaulLaux commented on src/note_encryption/domain.rs:
```diff
@@ -78,28 +72,32 @@ pub(super) fn prf_ock_orchard(
-        NOTE_VERSION_BYTE_V2 | NOTE_VERSION_BYTE_V3 => Some(*version),
-        _ => None,
-    })
+/// Checks if the version of the note plaintext matches the expected version for the domain.
+pub(super) fn validate_note_version<D: OrchardDomainCommon>(
+    plaintext: &D::CompactNotePlaintextBytes,
+) -> bool {
+    plaintext
+        .as_ref()
+        .first()
+        .map_or(false, |version| *version == D::NOTE_VERSION_BYTE)
```
We need to restore `parse_note_version`

Even if now we don't support other versions today, the encryption/decryption need to be version agnostic.

dmidem replied:
@PaulLaux , I still have a question.

Inside the `parse_note_plaintext_without_memo` function, we need to extract the asset from the plaintext. We know three facts: the Orchard flavor (Vanilla or ZSA), the compact note size, and the note version byte. Which criteria should we use to decide how to extract the asset: either use the native asset or extract it from the plaintext? We can't rely on the note version byte only and ignore the size of the plaintext, like we did before, to avoid out-of-bounds panics.

On the other hand, to extract the asset, we can ignore the version byte value and use the Orchard flavor or compact note size alone to decide how we extract the asset. This is the way I implemented it now (with additional note version validation, which has nothing to do with asset extraction).

Do you mean we need to use a combination of the note version byte and the flavor/plaitext size? Like this:

`let version = parse_note_version(plaintext.as_ref())?;
let asset = match version {
    NOTE_VERSION_BYTE_V2 if D::COMPACT_NOTE_SIZE == COMPACT_NOTE_SIZE_VANILLA => AssetBase::native(),
    NOTE_VERSION_BYTE_V3 if D::COMPACT_NOTE_SIZE == COMPACT_NOTE_SIZE_ZSA => {
        let bytes = plaintext.as_ref()[COMPACT_NOTE_SIZE_VANILLA..COMPACT_NOTE_SIZE_ZSA]
            .try_into()
            .unwrap();

        AssetBase::from_bytes(bytes).into()
    },
    _ => return None,
}
`
In this case, we don't need the new `extract_asset` methods of the OrchardDomainCommon I added. Or we can use  `extract_asset` this way:
`let version = parse_note_version(plaintext.as_ref())?;
let asset = match version {
    NOTE_VERSION_BYTE_V2 if D::COMPACT_NOTE_SIZE == COMPACT_NOTE_SIZE_VANILLA => D::extract_asset(plaintext)?,
    NOTE_VERSION_BYTE_V3 if D::COMPACT_NOTE_SIZE == COMPACT_NOTE_SIZE_ZSA => D::extract_asset(plaintext)?,
    _ => return None,
}
`

PaulLaux reviewed (commented):
One more fix is required, not "baking" the version byte into the domain

PaulLaux reviewed (approved):
Very nice
